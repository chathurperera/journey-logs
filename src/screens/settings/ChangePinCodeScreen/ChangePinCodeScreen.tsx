import { Icon } from '@rneui/base';
import React, { useEffect, useRef, useState } from 'react';
import { Image, View } from 'react-native';
import ReactNativePinView from 'react-native-pin-view';

import { images } from '@jl/assets';
import { Text } from '@jl/components';
import { tw } from '@jl/config';
import { Color, Route, TextAlignment, TextVariant } from '@jl/constants';
import { EncryptionService, NavigationService, NoteEncryption, ToastService } from '@jl/services';
import { useDispatch, useSelector } from '@jl/stores';

import { BaseScreenLayout } from '../../components/BaseScreenLayout';

export function ChangePinCodeScreen({ route }) {
  const pinView = useRef(null);
  const dispatch = useDispatch();

  const { masterKey } = route.params.params;
  const { userId } = useSelector(state => state.userStore.userData);
  const { salt } = useSelector(state => state.encryptionStore);

  const [showRemoveButton, setShowRemoveButton] = useState(false);
  const [enteredPin, setEnteredPin] = useState('');
  console.log('salt in ChangePinCodeScreen', salt);

  const generateNewRecoveryKey = async () => {
    try {
      const newRecoveryKey = await EncryptionService.generateNewEncryptedRecoveryKey(enteredPin, salt, masterKey);

      await NoteEncryption.savePinAndRecoveryKey(userId, salt, newRecoveryKey);
      dispatch.encryptionStore.setRecoveryKey(newRecoveryKey);

      ToastService.success('Success', 'PIN updated successfully');
      NavigationService.navigate(Route.SettingsTab);
    } catch (error) {
      ToastService.error('Error', 'Something went wrong');
    }
  };

  useEffect(() => {
    if (enteredPin.length > 0) {
      setShowRemoveButton(true);
    } else {
      setShowRemoveButton(false);
    }

    if (enteredPin.length === 4) {
      generateNewRecoveryKey();
    }
  }, [enteredPin]);

  const renderTitles = () => {
    return (
      <View style={tw`mt-12.5`}>
        <Text variant={TextVariant.Title2} textAlign={TextAlignment.Center}>
          Enter your New PIN
        </Text>
      </View>
    );
  };

  return (
    <BaseScreenLayout>
      <View style={tw`flex-1 justify-center`}>
        <View style={tw`flex-row justify-center`}>
          <Image source={images.logo} style={tw`w-20 h-20`} />
        </View>
        {renderTitles()}
        <ReactNativePinView
          inputSize={32}
          ref={pinView}
          pinLength={4}
          buttonViewStyle={tw`bg-[${Color.Neutral.JL50}]`}
          buttonTextStyle={tw`text-[${Color.Neutral.JL900}] text-4xlg font-normal`}
          buttonSize={70}
          buttonAreaStyle={tw`px-6`}
          inputAreaStyle={tw`mb-6`}
          inputViewStyle={tw`w-5 h-5`}
          inputViewEmptyStyle={tw`bg-[${Color.Neutral.JL50}]`}
          onValueChange={value => setEnteredPin(value)}
          onButtonPress={key => {
            if (key === 'custom_right') {
              pinView.current.clear();
            }
          }}
          //@ts-ignore
          customRightButton={
            showRemoveButton ? <Icon type="feather" name="delete" size={30} color={Color.Neutral.JL500} /> : undefined
          }
          style={tw`mt-10.75`}
        />
      </View>
    </BaseScreenLayout>
  );
}
