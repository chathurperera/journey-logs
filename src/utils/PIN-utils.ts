import moment from 'moment';

export const isPinSessionExpired = (lastSessionTimestamp: number, maxDurationInMinutes: number = 5) => {
  const timeDifference = moment().diff(moment(lastSessionTimestamp * 1000), 'minute');
  console.log('timeDifference', timeDifference);

  return timeDifference > maxDurationInMinutes;
};
