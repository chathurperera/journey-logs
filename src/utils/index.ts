export * from './moment-utils';
export * from './render-utils';
export * from './PIN-utils';
