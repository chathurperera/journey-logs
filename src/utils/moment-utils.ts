import moment from 'moment';

export const getCurrentTimestampInMilliSeconds = () => {
  const timestamp = moment().unix();
  return timestamp;
};

export const getRelativeTimeFromTimestamp = (timestampValue: number) => {
  const relativeTime = moment.unix(timestampValue).fromNow();
  return relativeTime;
};
