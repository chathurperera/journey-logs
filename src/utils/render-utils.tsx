import { NavigationContext } from '@react-navigation/native';
import { render } from '@testing-library/react-native';
import React, { ReactNode } from 'react';
import { Provider } from 'react-redux';

import { store } from '@jl/stores';

type RenderParams = Parameters<typeof render>;

export function renderWithProviders(ui: RenderParams[0], options?: RenderParams[1]) {
  const BaseProviders = ({ children }: { children: ReactNode }) => (
    <Provider store={store}>
      <NavigationContext.Provider
        //@ts-ignore
        value={{
          isFocused: () => true,
          addListener: jest.fn(() => jest.fn()),
        }}>
        {children}
      </NavigationContext.Provider>
    </Provider>
  );

  return render(ui, { wrapper: BaseProviders, ...options });
}
