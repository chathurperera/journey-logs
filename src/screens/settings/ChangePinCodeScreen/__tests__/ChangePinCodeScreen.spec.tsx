import React from 'react';

import { renderWithProviders } from '@jl/utils';

import { ChangePinCodeScreen } from '../ChangePinCodeScreen';

describe('<ChangePinCodeScreen />', () => {
  const testID = 'ChangePinCodeScreenTestID';
  const getRenderedScreen = () => renderWithProviders(<ChangePinCodeScreen testID={testID} />);

  it('should render ChangePINCodeScreen correctly', () => {
    const renderer = getRenderedScreen();
    const renderTree = renderer.toJSON();
    expect(renderTree).toMatchSnapshot();
  });

  it('should find the ChangePINCodeScreen via testID', () => {
    const { getByTestId } = getRenderedScreen();
    const foundScreen = getByTestId(testID);
    expect(foundScreen).toBeTruthy();
  });
});
