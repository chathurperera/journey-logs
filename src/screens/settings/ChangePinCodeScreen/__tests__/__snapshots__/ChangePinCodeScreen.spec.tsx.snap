// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChangePinCodeScreen /> should render ChangePINCodeScreen correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#ffffff",
      "height": "100%",
      "paddingTop": 50,
      "position": "relative",
    }
  }
  testID="ChangePinCodeScreenTestID"
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  />
</RCTSafeAreaView>
`;
